### DISCUSSION TITLE
How to make optional parameters to be automatically ignored in findOne `FilterQuery<T>` parameter?

### DISCUSSION THREAD

**Discussion by EvilCheetah** [mikro-orm/mikro-orm#4368] [Q&A]
I have a user entity, and in repository trying to find by several options: user_id, email and username. I have an interface that describes acceptable parameters for the function that utilizes `findOne` method. However, when I tried to pass it via the object, MikroORM, for some reason inserts undefined into the fields, which makes a repository method to return null.

What I currently have:
```ts
/// Entity: user.entity.ts
@Entity()
export class User
{
    @PrimaryKey()
    user_id:       number;

    @Property({ unique: true })
    username:      string;

    @Property({ unique: true })
    email:         string;

    @Property()
    password:      string;

    @Property({ nullable: true })
    refresh_token?: string;
}


/// Interface: UniqueCriteriaInterface
export interface UserUniqueCriteria
{
    user_id?:  number;

    email?:    string;

    username?: string;
}


/// Function(in NestJS Service): 
async find_one_by({ user_id, username, email }: UserUniqueCriteria)
{
    return this.userRepository.findOne({
        ...(user_id  ? { user_id }  : {}),
        ...(username ? { username } : {}),
        ...(email    ? { email }    : {})
    });
}
```
____
What I preferably  want to have:
```ts
/// Function(in NestJS Service): 
async find_one_by(criteria: UserUniqueCriteria)
{
    return this.userRepository.findOne(criteria);
}
```

**Comment by B4nan** [comment]
I think the question is more about where those `undefined` values come from, you should deal with them there instead of when calling the ORM methods. It can be done via one-liner, you don't need such destructing magic as in your example.

To me this is a feature, it protects you from doing things like `em.findOne(User, { id: undefined })`, which according to what you expect would return random user, I doubt you would want that. But maybe it's more confusing than helpful, and we should change it in v6. The protection for empty queries will be still there (e.g. doing `em.findOne(User, {})` will always fail).

**Reply by EvilCheetah** [reply]
Now it makes sense, but it seems like ORM tried to get users where, for instance, `email was supplied and other fields were undefined`. What what I can remember about SQL, they don't have undefined as a type nor as value, but `NULL`. That's why, I guess, we supply `nullable` property. 

Talking about `random`, I have reviewed documentation for SQLite, Postgres, MySQL, and it seems they are using the same functionality, either `RANDOM()` for [SQLite](https://www.sqlite.org/lang_corefunc.html#random) and [Postgres](https://www.postgresql.org/docs/current/functions-math.html), and `RAND()` for [MySQL](https://dev.mysql.com/doc/refman/5.7/en/mathematical-functions.html#function_rand):
```sql
SELECT * FROM ${TABLE_NAME}
   ORDER BY RANDOM() 
   LIMIT 1;
```
**Source**: [StackOverflow - Select random row from a sqlite table](https://stackoverflow.com/questions/2279706/select-random-row-from-a-sqlite-table)
___
For [MongoDB](https://www.mongodb.com/docs/manual/reference/operator/aggregation/sample/), it uses `$sample` query within `aggregate` method:
```ts
db.mycoll.aggregate([{ $sample: { size: 1 } }])
```
**Source**:  [StackOverflow - How can I get a random record from MongoDB?](https://stackoverflow.com/questions/2824157/how-can-i-get-a-random-record-from-mongodb)
___

In my case, I know for sure that there will be at least one parameter. Additionally, I can incorporate `class-validator` with ValidateIf decorator to ensure that at least one of the items was provided.

Thanks for the advice! You are amazin!
